[Spring MVC] 망고 step 5,6 미션 제출합니다.

llffmmgg opened the pull request:
안녕하세요, 커찬님! 반갑습니다~!!
2주 동안 잘 부탁드려요 🙇

#### 구현 시 고려했던 점

- 도메인 클래스에 setter를 사용하지 않기 위해 `BeanPropertyRowMapper` 대신 
`reservationRowMapper` 를 만들어 JSON을 객체에 매핑했습니다. 
- 최대한 불필요한 코드를 없애려고 노력했습니다.

부족하지만, 리뷰해 주시면 이를 바탕으로 공부하고 재리뷰 올리겠습니다!
감사합니다. 😊

llffmmgg committed:
chore: 의존성 추가

llffmmgg committed:
chore: application.properties 추가

llffmmgg committed:
test: 기본 테스크 코드 추가

llffmmgg committed:
feat: schema.sql 추가

llffmmgg committed:
jdbcTemplate 빈 추가

llffmmgg committed:
feat: getReservation(), jdbcTemplate 적용

llffmmgg committed:
test: 이전 단계 테스트 삭제

llffmmgg committed:
refactor: jdbcTemplate 빈 설정 파일 삭제

JdbcTemplate 은 설정으로 등록된 DataSource 스프링 빈을 사용하여 스프링 빈으로 등록되므로 삭제함

llffmmgg committed:
refactor: setter 제거를 위해 BeanPropertyRowMapper 대신 커스텀 RowMapper 적용

llffmmgg committed:
refactor: 불필요한 설정 제거

leegwichan commented on src/main/java/roomescape/domain/reservation/domain/Reservation.java:
```diff
@@ -5,6 +5,10 @@
 
 public class Reservation {
 
+    private Reservation() {
+
+    }
```
private 기본 생성자가 필요한 이유는 무엇인가요?

llffmmgg replied:
`BeanPropertyRowMapper`을 사용할 때 포함했었는 데 필요없어서 삭제할게요 !

leegwichan commented on src/main/java/roomescape/domain/reservation/domain/Reservation.java:
![image](https://github.com/user-attachments/assets/4ecefce6-f840-4284-b82f-d5770919aab1)
Reservation 이외에 다른 파일도 전부 확인해주세요!

leegwichan replied:
1. 현재 Reservation 파일에서 상수가 무엇인지, 인스턴스 변수가 무엇인지, 생성자가 무엇인지, 메서드가 무엇인지 각각 구분해서 말해주세요. 
2. 현재 같이 배치하게된 이유는 무엇인가요?

llffmmgg replied:
1.  
**인스턴스 변수**
```
private Long id;
private String name;
private LocalDate date;
private LocalTime time;
```
**생성자**
```
    public Reservation(final Long id, final String name, final LocalDate date, final LocalTime time) {
        this.id = id;
        this.name = name;
        this.date = date;
        this.time = time;
    }

    public Reservation(final String name, final LocalDate date, final LocalTime time) {
        this.name = name;
        this.date = date;
        this.time = time;
    }
```
**메서드**
```
    public Long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public LocalDate getDate() {
        return date;
    }

    public LocalTime getTime() {
        return time;
    }
```

2. 수정했었는데, `MissionStepTest `에서 reservation을 reservationResponse 타입으로 변경하던 도중 shift f6를 눌러 작업하다가 뭘 잘못눌렀는지, Reservation 클래스 명이 reservationResponse 로 바뀌고 해당 클래스 사용되었던 부분도 다 바뀌었더라구요 그래서 마지막 커밋을 취소하고 수정할 때 추가하는 걸 깜빡했었습니다. ㅜㅜ
앞으로는 확실히 확인해볼게요 !

leegwichan replied:
[필수 답변]

> 앞으로는 확실히 확인해볼게요 !

인스턴스 변수, 생성자, 메서드에 대해서 잘 알고 계시는데, 왜 위에 제가 언급한 순서대로 클래스를 작성하지 않으셨나요? 분명히 제안드리는 것과 다르게 해도 된다고 했지만, 그 이유를 명백히 남겨달라고 말씀드렸습니다.

> final 키워드를 사용할 때 단순히 코드를 작성하면서 재할당되나? 생각하고 넣는 식으로 작성했던 것 같아요.
나름대로 찾아보며 기준을 세워봤어요..! 피드백 부탁드립니다.
[기준] 클래스 내 불변인 멤버 필드와 메서드 내부에서 값 또는 인스턴스가 변하지 않는 파라미터에 붙임

인스턴스 변수
```java
    private final Long id;
    private String name;
    private LocalDate date;
    private LocalTime time;
```

???????????????????????????????????????????????????????????????????????????????????????????????
???????????????????????????????????????????????????????????????????????????????????????????????
???????????????????????????????????????????????????????????????????????????????????????????????

llffmmgg replied:
> 인스턴스 변수, 생성자, 메서드에 대해서 잘 알고 계시는데, 왜 위에 제가 언급한 순서대로 클래스를 작성하지 않으셨나요? 분명히 제안드리는 것과 다르게 해도 된다고 했지만, 그 이유를 명백히 남겨달라고 말씀드렸습니다.

아.. 진짜 많이 확인했는데, 뭐가 씌였는지 계속 생성자를 밑에 뒀었네요.. 저도 다시 보는데 왜 그랬는지 이해가 안 되네요. 수정했습니다!

 >인스턴스 변수

    private final Long id;
    private String name;
    private LocalDate date;
    private LocalTime time;

인스턴스 변수` id`에만 `final`을 붙힌 이유는 name, date, time 필드 애플리케이션의 요구 사항에따라 변경의 가능성이 있어서 id에만 final을 붙였어요 ! (제가 코멘트를 제대로 이해하고 답변한게 맞을까요 ?)

leegwichan replied:
> 인스턴스 변수 id에만 final을 붙힌 이유는 name, date, time 필드 애플리케이션의 요구 사항에따라 변경의 가능성이 있어서 id에만 final을 붙였어요 !

만약에 final을 제거하는 작업이 리소스가 크게 든다면 망고님 의견처럼 미리 대비할 것 같아요. 하지만 저는 그렇게 리소스가 많이 들지 않는다고 생각해서 붙이는 편입니다.

leegwichan commented on src/main/java/roomescape/domain/reservation/repository/ReservationRepository.java:
```diff
@@ -26,6 +36,16 @@ public boolean removeReservation(final long id) {
     }
 
     public List<Reservation> getReservations() {
-        return reservations;
+        final String sql = "SELECT * FROM reservation";
+
+        return jdbcTemplate.query(sql, reservationRowMapper);
     }
+
+    private final RowMapper<Reservation> reservationRowMapper = (resultSet, rowNum) ->
+            new Reservation(
+                    resultSet.getLong("id"),
+                    resultSet.getString("name"),
+                    resultSet.getObject("date", LocalDate.class),
+                    resultSet.getObject("time", LocalTime.class)
+            );
 }
```
망고가 final을 붙이는 기준은 무엇인가요?

llffmmgg replied:
값을 초기화후 변경하지 않는 것이 확실하면 붙이는 것 같아요 !

leegwichan replied:
제가 망고 코드를 봤을때는 그렇지 않은 부분이 너무 많은 것 같은데요?
어디는 final이 메서드 인자에 붙어 있을 때도 있고, 없을 때도 있고
어디는 final이 메서드 내 변수에 붙어 있을 때도 있고, 없을 때고 있고
어디는 final이 인스턴스 변수에 붙어 있을 때도 있고, 없을 때고 있고
명확한 기준을 세워서 모두 붙이거나 떼거나 통일해주세요. 만약에 지금처럼 명확한 기준이 없다면 코드 내에서 final이 주는 의미가 없어질 것 같네요.
```java
public class Reservation {
    private Long id;
    private String name;
    private LocalDate date;
    private LocalTime time;
    ...
}
```
```java
@Repository
public class ReservationRepository {

    private final JdbcTemplate jdbcTemplate;
    private List<Reservation> reservations = new ArrayList<>();
    private AtomicLong index = new AtomicLong(1);
    ...
}
```
```java
    @PostMapping("/reservations")
    public ResponseEntity<ReservationResponse> createReservation(
            @RequestBody final ReservationRequest reservationRequest) {
        ReservationResponse result = reservationService.createReservation(reservationRequest);
        URI Location = URI.create("/reservations/" + result.id());

        return ResponseEntity.created(Location).body(result);
    }
```
```java
    @Test
    void getReservations_테스트() {
        List<Reservation> reservations = reservationRepository.getReservations();
        int beforeCount = reservations.size();

        jdbcTemplate.update("INSERT INTO reservation (name, date, time) VALUES (?, ?, ?)", "브라운", "2023-08-05",
                "15:40");
        reservations = reservationRepository.getReservations();
        int afterCount = reservations.size();

        Assertions.assertThat(beforeCount + 1).isEqualTo(afterCount);
    }
```

llffmmgg replied:
`final` 키워드를 사용할 때 단순히 코드를 작성하면서 재할당되나? 생각하고 넣는 식으로 작성했던 것 같아요.

나름대로 찾아보며 기준을 세워봤어요..! 피드백 부탁드립니다.
[기준]
 클래스 내 불변인 멤버 필드와 메서드 내부에서 값 또는 인스턴스가 변하지 않는 파라미터에 붙임
> 객체 생성 이후 값이 변하지 않음을 명확히 표현하여 코드의 안정성과 예측 가능성을 높임.
> 해당 변수가 변경되지 않음을 명확히 전달하기 위함

지역 변수는 함수 내부에서 파라미터 값을 받아 원하는 값을 만들기위한 것이므로 final을 강제하지는 않는다.
(함수라는 관점에서) 필요하다고 생각하면 명시!
좋은 피드백 감사합니다 ! 🙇‍♂️

leegwichan replied:
[필수 답변]

> (함수라는 관점에서) 필요하다고 생각하면 명시!

이런 것도 모호한 표현이 될 수 있어요. 지금은 지역 변수에서 final을 붙인 경우가 없네요. 그러면 "필요하다고 생각"하는 예시 하나만 작성하실 수 있나요? 근거와 함께 코멘트 남겨주세요!

llffmmgg replied:
`final String sql = "select * from reservation` 에 추가했었습니다. 
해당 sql 문은 초기화 된 후 `jdbcTemplate`를 통해 실행되기 전까지 변경되면 안된다고 생각하여 final을 추가했어요 !

이 코멘트를 받고 다시 생각해봤는데, 지역 변수에 강조의 표시로 final을 쓰는 기준을 사용하면 협업자가 final 붙은 변수에 대한 이해는 높아져도 다른 변수들에는 변경될 수 있나? 없나 ? 오히려 헷갈리게 만드는 것 같기도 합니다..! 

해당 미션에서 지역 변수에서는 다 빼도록 할게요 !

leegwichan commented on src/main/resources/schema.sql:
EOF 에러가 일어날 수 항상 주의해주세요!

llffmmgg replied:
주의하겠습니다 ~!

leegwichan commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE reservation
+(
+    id      BIGINT       NOT NULL AUTO_INCREMENT,
+    name    VARCHAR(255) NOT NULL,
+    date    VARCHAR(255) NOT NULL,
+    time    VARCHAR(255) NOT NULL,
+    PRIMARY KEY (id)
+);
```
1. H2 데이터베이스의 자료형에는 어떤 것들이 있나요?
2. 각 필드가 해당 자료형에 어울린다고 생각하시나요?

llffmmgg replied:
1. 정수형
- TINYINT, SMALLINT, INTEGER, BIGINT
2. 실수형
- REAL, DOUBLE, DECIMAL, NUMERIC
3. 문자열형
- CHAR(n), VARCHAR(n), CLOB
4. 날짜 및 시간형
- DATE, TIME, TIMESTAMP

id와 name은 적절해 보이며, name의 크기는 줄이는 것이 좋을 것 같습니다!
DATE와 TIME은 숫자로 저장되므로, 동일한 시간을 저장할 때 VARCHAR로 저장하는 것보다 공간적으로 효율적입니다.
또한, 비교 및 정렬 속도 또한 더 빠르므로 DATE와 TIME을 사용하는 것이 좋습니다. 따라서 수정하겠습니다!

leegwichan commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -2,70 +2,52 @@
                 .then().log().all()
-                .statusCode(200)
-                .body("size()", is(1));
+                .statusCode(200).extract()
+                .jsonPath().getList(".", Reservation.class);
```
(테스트는 잘 통과 하겠습니다만,) 해당 값을 Reservation으로 매핑하는 게 맞을까요? 망고 코드에 맞춰 조금 커스텀 해볼까요?

llffmmgg replied:
매핑 값을 해당 컨트롤러("`/reservations`")의 반환값과 동일하게 변경했습니다 !

leegwichan commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -2,70 +2,52 @@
 
 import io.restassured.RestAssured;
 import io.restassured.http.ContentType;
+import java.sql.Connection;
+import java.sql.SQLException;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.jdbc.core.JdbcTemplate;
 import org.springframework.test.annotation.DirtiesContext;
+import roomescape.domain.reservation.domain.Reservation;
+
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.hamcrest.CoreMatchers.*;
```
Google Java Style Convention 중

> 3.3 Import statements
3.3.1 No wildcard imports
Wildcard imports, static or otherwise, are not used.
3.3.3 Ordering and spacing
Imports are ordered as follows:

> - All static imports in a single block.
> - All non-static imports in a single block.

> If there are both static and non-static imports, a single blank line separates the two blocks. There are no other blank lines between import statements.

leegwichan commented on src/main/resources/application.properties:
properties 파일 말고 다른 형식으로 설정하는 방법이 있을까요?

llffmmgg replied:
yml 파일을 사용해서 설정할 수 있습니다. 
-가독성이 좋고 계층적으로 구조를 설정할 수 있다는 장점이 있어요 !

leegwichan commented on src/main/java/roomescape/domain/reservation/repository/ReservationRepository.java:
```diff
@@ -26,6 +36,16 @@ public boolean removeReservation(final long id) {
     }
 
     public List<Reservation> getReservations() {
-        return reservations;
+        final String sql = "SELECT * FROM reservation";
+
+        return jdbcTemplate.query(sql, reservationRowMapper);
     }
+
+    private final RowMapper<Reservation> reservationRowMapper = (resultSet, rowNum) ->
+            new Reservation(
+                    resultSet.getLong("id"),
+                    resultSet.getString("name"),
+                    resultSet.getObject("date", LocalDate.class),
+                    resultSet.getObject("time", LocalTime.class)
+            );
 }
```
아래의 말이 참일지 거짓일지 생각해보죠. (인터넷 검색하지 말고 망고의 의견을 달아보도록 하죠)

1. 1000번의 GET 요청을 날리면, `"SELECT * FROM reservation"` String 객체가 1000개 생긴다.
2. 1000번의 GET 요청을 날리면, `reservationRowMapper`(RowMapper<Reservation>) 객체가 1000개 생긴다.
3. 1000번의 GET 요청을 날리면, `ReservationRepository`, `JdbcTemplate` 객체가 각각 1000개 생긴다.

llffmmgg replied:
1번 거짓, 한번만 생성되고 다음부터는 이미 생성된 문자열을 참조합니다 ! 
3번 거짓, ReservationRepository, JdbcTemplate은 Spring Bean으로 하나만 존재합니다. 
2번  ReservationRepository가 하나의 객체만 생성되기 때문에 1개만 생성될 것 같아요..!

leegwichan replied:
> 1번 거짓, 한번만 생성되고 다음부터는 이미 생성된 문자열을 참조합니다 !

당연히 제가 정답만을 기대하는 건 아닌거 아시죠? 분명히 메서드 내에서 새로 객체를 만들어서 할당하는데, 왜 String 객체가 1000개가 생성되지 않나요? 이상하지 않나요?

> 3번 거짓, ReservationRepository, JdbcTemplate은 Spring Bean으로 하나만 존재합니다.
2번 ReservationRepository가 하나의 객체만 생성되기 때문에 1개만 생성될 것 같아요..!

2번에 확신이 없으신 건 아니죠? 물론 3번이 거짓인 것 처럼 `RowMapper`는 한 개만 생성됩니다.
그렇다면 `RowMapper`를 `private static final`이 아닌 `private final`로 선언하신 이유가 있으신가요?

llffmmgg replied:
> 당연히 제가 정답만을 기대하는 건 아닌거 아시죠? 분명히 메서드 내에서 새로 객체를 만들어서 할당하는데, 왜 String 객체가 1000개가 생성되지 않나요? 이상하지 않나요?

`String`은 매우 빈번하게 사용되는 타입입니다. 따라서 `String Constant Pool` 이라는 공간을 Heap 영역에 만들어 매번 새로운 객체를 생성하지 않고 같은 String이면 해당 객체를 참조하는 형태로 사용합니다. 
추가적으로 `String`은 불변 객체이기 때문에 참조하던 값을 변경하여도 값이 다른데 같은 참조를 가지는 일을 방지할 수 있습니다 ! 
| new를 사용해서 할당할 경우 중복되는 String이어도 Heap 영역의 다 다른 공간에 저장된다고 합니다 !

> 2번에 확신이 없으신 건 아니죠? 물론 3번이 거짓인 것 처럼 RowMapper는 한 개만 생성됩니다.
그렇다면 RowMapper를 private static final이 아닌 private final로 선언하신 이유가 있으신가요?

`ReservationRepository`의 인스턴스가 하나 존재하기 때문에 RowMapper 객체도 하나만 생성되는 것이 당연합니다. 
추가적으로 `private final` 로 사용한 이유는 우선 `static`을 사용하면 인스턴스와 상관없이 사용 가능하고 인스턴스 개수와 상관없이 한 개가 존재한다는 장점이 있습니다. 
`RowMapper`는 ResultSet을 `reservation `으로 매핑하고 있습니다.  따라서 `ReservationRepository`에서만 쓰일 가능성이 높기때문에 `private`로 선언했습니다. 또한 `ReservationRepository` 자체가 한 개만 존재하기 때문에 static을 추가하지 않았습니다 !

leegwichan replied:
> ReservationRepository 자체가 한 개만 존재하기 때문에 static을 추가하지 않았습니다 !

 그러면 static final로 했을 때 장점을 생각해보면... 테스트 상황에서도 RowMapper가 한 개만 생성될 수 있겠네요. 각각의 테스트에서 ReservationRepository를 하나씩 만들어 사용한다고 가정해보죠. 그러면 인스턴스 변수로 하면 인스턴수 개수만큼 생성이 되고, 상수(클래스 변수)로 생성하면 1개만 생성이 되겠네요.

 저는 Spring의 컨텍스트와 상관없이 자바의 최적화 방안이 있다면 그것을 따르는 편입니다. 왜냐면 결국에는 Spring은 Java 기반으로 만들어졌기 때문에 Java에 해당하는 규칙을 지키면 Spring에서는 자동으로 지켜집니다. 하지만 Spring의 최적화 방안에 따라 생각하면 그 이외의 상황에서 최적화가 안될 수 있어요. (ex. 테스트, ...) 이것은 하나의 제안이나 망고가 인스턴스 변수로 사용하는 명백한 이유가 있다고 한다면 굳이 안바꾸셔도 좋을 것 같습니다.

llffmmgg replied:
테스트를 구현하는 관점과 최적화 레벨 관점에서 생각해보았을 때 커찬님 코멘트에 동의합니다 !

굳이 생각해보면 해당 변수의 생명 주기가 인스턴스보다 길다는게 문제라면 문제인데(불필요한 메모리), 해당 클래스는 싱글톤 bean 이므로 생명주기가 컨테이너와 큰 차이가 없어 상관없어보입니다..! 
`static final`로 수정할게요 ! 

최적화에 대해 단순하게 생각했었는 데, 이번 리뷰보고 자바 단계에서의 최적화 스프링 단계에서의 최적화 나눠 생각할 수 있게 된 것 같아요 ! 감사합니당

leegwichan commented on src/main/java/roomescape/domain/reservation/repository/ReservationRepository.java:
저희 '클린 코드' 요구 사항 중에 한 줄을 가져오죠

> 테스트 도구를 활용해 본인이 정리한 기능 목록이 정상 동작함을 테스트 코드로 확인한다.

아래 ReservationRepository의 `getReservations()`의 정상 작동을 확인하려면 어떻게 해야 하나요? 직접 Application을 실행 시켜서 GET 요청을 보내서 확인해봐야겠죠. 하지만 해당 방법은 코드를 변경하기 힘든 구조가 됩니다. 다른 개발자가 함부로 해당 코드를 변경하지 못하겠죠 (테스트 하려면 복잡한 절차를 거쳐야 하니...)

그래서 ReservationRepository의 `getReservations()`가 정상 작동하는지 확인하는 테스트 코드를 작성해 보도록 하죠. 어려우실 수 있지만, 꼭 필요한 과정이니 테스트 할 수 있는 방법을 잘 찾아보시기 바랍니다~

llffmmgg replied:
일단 검색없이 혼자 구현해봤어요 ! 피드백 부탁드립니다..!

leegwichan commented on src/main/java/roomescape/domain/reservation/repository/ReservationRepository.java:
![image](https://github.com/user-attachments/assets/5d6a4107-cea3-49ee-94e6-d9712928cc67)
전반적으로 모든 코드를 한 번 둘러보시는 걸 추천드립니다.

llffmmgg replied:
넵 반영하겠습니다.

leegwichan reviewed (changes requested):
안녕하세요 망고! 커찬입니다. 리뷰어로 만나뵙게 되어 반갑습니다.
2주라는 시간이 있고 진도가 빠르시니 고봉밥 리뷰 남겨봤습니다~ 해당 코멘트에 대해 깊이 생각해보시고 리뷰 반영&코멘트 부탁드리도록 할께요!

> 도메인 클래스에 setter를 사용하지 않기 위해 BeanPropertyRowMapper 대신
reservationRowMapper 를 만들어 JSON을 객체에 매핑했습니다.

(진짜 몰라서 물어보는데...) `BeanPropertyRowMapper`가 무엇인가요?

> 최대한 불필요한 코드를 없애려고 노력했습니다.

불필요한 코드가 어떤 것이 있었고, 어떻게 노력하셨나요? 어떤 점에 신경 써주셨는지 구체적으로 이야기 해주셔야 좋은 피드백이 가능할 것 같아요!

llffmmgg committed:
eof, 코딩 컨벤션 맞춤

llffmmgg committed:
test: reservationRepository, getReservation() 테스트 코드 작성

llffmmgg reviewed (commented):
안녕하세요, 커찬님!

제가 설명이 좀 부족했던 것 같아요. 
`BeanPropertyRowMapper`은 스프링 JDBC에서 제공하는 `RowMapper` 의 구현체로 SQL의 결과(ResultSet)를 지정한 객체로 자동 매핑해주는 기능을 가지고 있습니다. 
예) 
`new BeanPropertyRowMapper<>(Reservation.class)
`
기본적으로 빈 생성자와 setter를 이용한다고 알고있어 setter를 사용하지 않기 위해 사용하지 않았습니다. (private 필드가 있는 경우)

불필요한 코드는, 예를 들어 ResponseEntity를 사용하고 있을 때 @ResponseBody를 붙이지 않는 것과 같은 것들입니다. 이번 과제에서는 저 정도 부분밖에 없었던 것 같아요 !
리뷰도 반영했어요.

고봉밥 리뷰 감사합니다. ㅎㅎ

llffmmgg committed:
refactor: 테스트 반환 값 변경 및 코딩 컨벤션

llffmmgg committed:
feat: if not exist 추가

llffmmgg committed:
refactor: 필요없는 코드 삭제

leegwichan commented on src/main/java/roomescape/domain/reservation/repository/ReservationRepository.java:
```diff
@@ -26,6 +35,16 @@ public boolean removeReservation(final long id) {
     }
 
     public List<Reservation> getReservations() {
-        return reservations;
+        final String sql = "SELECT * FROM reservation";
+
+        return jdbcTemplate.query(sql, reservationRowMapper);
     }
+
+    private final RowMapper<Reservation> reservationRowMapper = (resultSet, rowNum) ->
+            new Reservation(
+                    resultSet.getLong("id"),
+                    resultSet.getString("name"),
+                    resultSet.getObject("date", LocalDate.class),
+                    resultSet.getObject("time", LocalTime.class)
+            );
 }
```
분명히 한 번 언급드리고 모든 파일 확인해달라고 했는데... `RowMapper<Reservation>`는 아래 규칙을 안지킨 이유가 있나요?
(컨벤션을 지키는 건 생각보다 중요한 문제입니다. 생각한 객체가 생각한 위치에 없다면, 사용자마다 원하는 객체를 아무대나 만들게 되겠죠? 그러면 유지보수가 매우 힘들어질 겁니다. 개인적으로는 한 번 언급한 내용에 대해 다시 언급하는 일은 없었으면 좋겠습니다.)
![image](https://github.com/user-attachments/assets/1d4a8eef-c172-4697-a2a0-b177464e52c5)

llffmmgg replied:
`RowMapper<Reservation>` 같은 람다식을 사용한 형태를 잘 사용해보지 않아서 착각했습니다.. 
피드백을 더 잘보고 반영한 후 리뷰드릴게요 죄송합니다.

컨벤션의 중요성도 이해했습니다 감사합니다 !

leegwichan commented on src/test/java/roomescape/domain/reservation/repository/ReservationRepositoryTest.java:
코멘트 드린 대로 ReservationRepository에 테스트를 만드셨네요 👍 그러면 이제 검색을 통해 최적화 해볼까요?

- `@SpringBootTest`를 이용하면 어떤 환경에서 Test를 하게 되나요? 만약 Bean들을 등록하게 된다면 너무 비효율적이지 않나요?
- Repository만 테스트할 수 있는 좋은 방법이 있을까요?

llffmmgg replied:
> @SpringBootTest를 이용하면 어떤 환경에서 Test를 하게 되나요? 만약 Bean들을 등록하게 된다면 너무 비효율적이지 않나요?

보통 통합 테스트를 할 때 사용합니다. 
`@SpringBootTest` 를 사용하면,  `webEnvironment` 를 통해 실행 환경을 설정할 수 있습니다 !
- MOCK
서버를 실행하지 않고(Mock 환경) 웹 컨텍스트만 로드합니다. 즉 ServletContainer인 내장 톰캣이 실행되지 않습니다 !
- RANDOM_PORT
실제 웹 환경을 제공하며 내장 서버도 실행되어 사용하지 않는 PORT를 랜덤으로 listen 상태로 유지합니다. ( 클라이언트의 요청을 기다리는 상태)
- DEFINED_PORT
랜덤포트와 같으며 지정한 PORT를 listen 합니다. (default port = 8080)
- NONE
웹 관련 설정을 사용하지 않습니다. 즉, 내장 톰캣을 실행하지 않으며 웹 환경을 포함하지 않는 테스트 환경을 구성할 때 사용됩니다.

DEFAULT 는 MOCK 이므로 현재는 서버를 실행하지 않고(Mock 환경) 웹 컨텍스트만 로드한 환경에서 테스트가 진행됩니다.

> Repository만 테스트할 수 있는 좋은 방법이 있을까요?

찾아보니 `@JdbcTest` 라는 것이 있어서 수정했습니다 ! 
`@JdbcTest `는 Jdbc 관련 설정만 적용됩니다. 
- 모든 스프링 Bean을 등록하지 않고 Jdbc 관련 Bean만 등록합니다. 
- In-memory DB를 사용합니다. 
- 각각의 테스트는 하나의 트랜잭션이 적용되고, 테스트가 끝나면 roll back 됩니다.

leegwichan commented on src/test/java/roomescape/domain/reservation/repository/ReservationRepositoryTest.java:
```diff
@@ -0,0 +1,32 @@
+    @Autowired
+    private ReservationRepository reservationRepository;
+
+    @Test
+    void getReservations_테스트() {
+        List<Reservation> reservations = reservationRepository.getReservations();
+        int beforeCount = reservations.size();
+
+        jdbcTemplate.update("INSERT INTO reservation (name, date, time) VALUES (?, ?, ?)", "브라운", "2023-08-05",
+                "15:40");
+        reservations = reservationRepository.getReservations();
+        int afterCount = reservations.size();
+
+        Assertions.assertThat(beforeCount + 1).isEqualTo(afterCount);
+    }
```
테스트의 의도가 무엇인지 파악하기 어렵네요. 
- beforeCount를 가져오는 이유는 무엇인가요?
- "두 번 호출해서 개수가 한 개 증가한다"라는 것으로 '전체 예약 조회'가 정상적으로 작동한다고 말 할 수 있나요?

llffmmgg replied:
> beforeCount를 가져오는 이유는 무엇인가요? 

처음 `@SpringBootTest ` 를 사용해서 repository test를 구현하고 모든 테스트를 한번에 돌릴 때, **DB에 테이블이 초기화가 되어 있는데 DDL을 다시 재실행해서 발생하는 오류가 발생했어요**. 
그래서 단순히 "테스트를 한번에 돌리면 같은 디비를 사용하네" 라고 생각해서 MissionStepTest에서 저장되는 데이터가 있을테니 `getReservatoion()`결과 
 카운트해서 `getReservations()`를 검증해야겠다고 생각해서 저런 형태의 이상한 테스트 코드를 작성했었어요 ..!

  >"두 번 호출해서 개수가 한 개 증가한다"라는 것으로 '전체 예약 조회'가 정상적으로 작동한다고 말 할 수 있나요? 

아니요 ! (위 답변에 저렇게 작성한 변명이 있습니다 ㅜㅜ)
`@JdbcTest`를 찾아보면서  `@SpringBootTest`는 기존 애플리케이션에 설정한 db를 사용하고  `@JdbcTest`는 인메모리 db를 (보통 h2, 설정도 가능) 사용한다는 것을 알았습니다.
일단 기본적으로 기존 애플리케이션에서도 h2-db를 사용해서 위에서 생각한 문제가 발생하지 않을 뿐만 아니라(db의 생명주기가 애플리케이션과 같기때문) `@JdbcTest`에서 실행되는 db는  `url='jdbc:h2:mem:9ca40b50-9792-4d2e-9ec0-323bb730a89b`  이고, 애초에 기존 애플리케이션과는 다른 db에서 테스트가 진행 중 이었기 때문에 해당 테이블이 비어있다는 것을 인지한 채로 테스트를 다시 작성했어요 !

leegwichan replied:
앞으로는 이렇게 고민되는 점이 있다면 PR 작성하시면서 같이 코멘트 남겨주시면 좋을 것 같아요! 그러면 제가 빠르게 팁을 드릴 수 있을 것 같습니다!

llffmmgg replied:
네! 앞으로 PR 작성할 때 고민되는 점이 있다면 같이 코멘트로 남기겠습니다.
피드백을 빠르게 받을 수 있어서 더 좋은 방향으로 개선할 수 있을 것 같아요. 감사합니다!

leegwichan commented on src/main/java/roomescape/domain/reservation/controller/ReservationController.java:
```diff
@@ -41,15 +41,16 @@ public ResponseEntity<List<ReservationResponse>> getReservations() {
 
     @DeleteMapping("/reservations/{reservationId}")
-    public ResponseEntity deleteReservation(@Validated  @PathVariable Long reservationId) {
+    public ResponseEntity deleteReservation(@Validated @PathVariable final Long reservationId) {
```
- `@Validated`는 무엇을 검증하는 건가요?
- reservationId가 `long`이 아닌 `Long`인 이유는 무엇인가요?

llffmmgg replied:
1. 처음 reservationId에 대한 검증을 처리할 방법을 생각해보면서 이것저것 해보다가 빼두질 않았네요..

`@Validated`는 `@RequestBody`,` @RequestParam`, `@ModelAttribute` 등을 검증하고, PathVariable은 검증하지 않습니다. 삭제하겠습니다 !

2. `Long`을 써야 할 이유가 없어 보이네요. `long`은 기본형(primitive type)이기 때문에 null을 허용하지 않으며, reservationId도 null이 들어오면 안 되므로 `long`을 사용하는 것이 적절해보이네요 ! 수정하겠습니다.

leegwichan reviewed (changes requested):
안녕하세요~ 커찬입니다.

 리뷰에서 언급드렸던 부분인데도 불구하고 반영하지 않은 부분이 있는 것 같아요. 반영하지 않은 이유가 있다면 명확하게 코멘트를 남겨주셔야 어떤 생각을 하는지 알고 좋은 피드백을 주고 받을 수 있다고 생각합니다. 그냥 반영하겠다고 말만하고 반영 안되어있는 부분이 있다면, 리뷰어는 생각보다 기분이 나쁘답니다~

 추가적으로 컨벤션이 잘 지켜져야 우리가 다루고자 할 부분에 더 집중해서 리뷰를 주고받을 수 있다고 생각해요. 지금도 더 언급하고 싶은 부분이 있지만 분량이 많아질 것 같아 다음으로 넘긴 주제들이 꽤 있습니다. 그러니 좀 더 컨벤션에 신경 써주시면 감사하겠습니다.

leegwichan commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -1,71 +1,50 @@
+import roomescape.domain.reservation.dto.ReservationResponse;
 
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
 @DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
```
[선택 답변]

지금 경우에는 Application이 돌아가는 상황에서 테스트를 돌리면 테스트가 실패해요~
이걸 해결할 수 있는 방법이 있을까요?

llffmmgg replied:
저번 코멘트에서 `WebEnvironment`에 대해 학습한 내용을 바탕으로 포트 번호를 변경해 보았는데,
테스트 API 요청이 여전히 8080번 포트로 가면서 에러가 발생하는 문제를 발견했습니다.
이를 해결하기 위해` @LocalServerPort` 애노테이션을 사용하여 랜덤 포트 번호를 가져온 뒤,
해당 포트를 `RestAssured.port`에 테스트 전에 설정해 주어 문제를 해결할 수 있었습니다!

leegwichan commented on src/main/java/roomescape/domain/reservation/controller/ReservationController.java:
[선택 답변]

여기에서 API가 추가되면 어떻게 관리하실 예정이신가요?

llffmmgg replied:
저는 도메인 별 [ 예) member, reservation, admin , auth ] 관련된 기능으로 컨트롤러를 분리합니다. 
그래서 reservation 관련 기능이면 그대로 해당 클래스에 추가하고 다른 도메인 관련 기능이 추가되면 새로운 클래스를 만들어 관리할 것 같아요. 

> 제가 생각하는 도메인

서비스 내 핵심 개념과 기능을 정의하는 영역이라고 생각합니다. 예를 들어 로그인(인증,인가), 예약 관련 기능, 멤버 관련 기능.. 

제가 리뷰의 핀트를 못잡고있다는 생각도 들어서 그런 상황이면 코멘트 달아주시면 다음 pr에 다시 생각해서 작성해볼게요 !

leegwichan replied:
> reservation 관련 기능이면 그대로 해당 클래스에 추가하고 다른 도메인 관련 기능이 추가되면 새로운 클래스를 만들어 관리할 것 같아요.

👍 👍 👍

개인적으로 제 생각을 덧붙이자면 뷰를 보여주는 것과 JSON을 보여준 것을 분리하는 것이 나중에 유지보수하기 편하다고 생각합니다. 그 return 형식에 따라 역할이 달라진다고 생각하거든요.

- HTML : (현재는) 단순한 HTML 파일을 반환하는 책임을 집니다.
- JSON : DB에 있는 결과를 토대로 올바른 행동을 한 후에 형식에 맞는 결과를 제공하는 책임을 집니다.

그래서 위와 같이 이유로 만약에 저였다면 `home()`. `reservation()` 컨트롤러를 분리할 것 같네요.

llffmmgg committed:
refactor: 컨트롤러, 서비스 매개변수 타입 변경

Long -> long

llffmmgg committed:
refactor: final 키워드 관련 수정

기준: 클래스 내 불변인 멤버 필드와 메서드 내부에서 값 또는 인스턴스가 변하지 않는 파라미터에 붙임
지역 변수는 함수 내부에서 파라미터 값을 받아 원하는 값을 만들기위한 것이므로 final을 강제하지는 않는다.

llffmmgg committed:
refactor: 지역 변수에 final 키워드 추가

지역 변수 내 final 키워드는 선택이다. 파라미터로 전달되기 전까지 변경되지 않음을 강조하기 위해 포함

llffmmgg reviewed (commented):
완벽히 하지 않던 공부 습관이 들어나는 것 같아요. 시간 써서 리뷰해주시는 데, 서로 밀도있는 리뷰가 되도록 더 노력하겠습니다. 
자세한 코멘트 감사합니다 !

llffmmgg committed:
refactor: ReservationRepositoryTest-> 단위테스트로 변경

llffmmgg committed:
chore: 필요없는 import 삭제

leegwichan commented on src/test/java/roomescape/domain/reservation/repository/ReservationRepositoryTest.java:
[필수 답변]

> refactor: ReservationRepositoryTest-> 단위테스트로 변경

유독 한 커밋 메시지가 눈에 띄네요. 단위 테스트는 무엇인가요?

llffmmgg replied:
> 단위 테스트

하나의 기능 또는 메서드를 기준으로 독립적으로 진행되는 가장 단위의 테스트입니다. 
- 새로운 기능 추가 시 수시로 빠르게 테스트 할 수 있다.
- 리팩토링 시에 안정성을 확보할 수 있다. ->(이건 프리코스하면서 직접 느끼었던 점입니다!)
- 테스팅에 대한 시간과 비용을 절감할 수 있다.
- 코드에 대한 문서가 될 수 있다.
이러한 장점들이 있어 실무에서는 단위를 선호하고 있다고 해요 !
 
> 통합 테스트 

기능 또는 메서드(모듈)를 통합하는 과정에서 모듈 간의 호환성을 확인하기 위해 수행되는 테스트입니다.
독립적인 기능에 대한 테스트가 아니라  웹 페이지로부터 API를 호출하여 올바르게 동작하는 지를 확인하는 것입니다. 

커밋 전 테스트는 테스트 목적은 단위테스트인데, @SpringBootTest 를 사용하고 있어 불필요하게 통합 테스트처럼 불필요한 bean을 등록하고 내장 서버도 실행하고 있어 `ReservationRepositoryTest-> 단위테스트로 변경` 이라는 커밋 메세지를 작성했어요!

leegwichan replied:
> 새로운 기능 추가 시 수시로 빠르게 테스트 할 수 있다.
리팩토링 시에 안정성을 확보할 수 있다. ->(이건 프리코스하면서 직접 느끼었던 점입니다!)
테스팅에 대한 시간과 비용을 절감할 수 있다.
코드에 대한 문서가 될 수 있다.
이러한 장점들이 있어 실무에서는 단위를 선호하고 있다고 해요 !

이것은 테스트의 장점을 말하고 있는 것이 아닌가 싶네요. 단위 테스트가 아니더라도 통합 테스트도 같은 장점을 가질 것 같아요.

제 의도는 "단위 테스트"는 무엇일까? 그러면 다른 테스트와 무슨 차이가 있을까?를 고민해보셨으면 좋겠어서 질문 드렸습니다. 그래서 아래 래퍼런스 한 번 참고해보시면 좋을 것 같아요!

https://tecoble.techcourse.co.kr/post/2021-05-25-unit-test-vs-integration-test-vs-acceptance-test/

leegwichan reviewed (changes requested):
이번에 Approve 드리고 싶었는데 시간도 많이 남았으니 한 번만 너 이야기 나눠보도록 하죠. [필수 답변]과 [선택 답변]을 나누어 놓았으니 [필수 답변]에 대해서는 이번에 답변 달아주시고  [선택 답변]은 꼭 이번이 아니라 step 7 진행하면서 남겨주셔도 상관 없습니다. 코멘트&리뷰 반영 하고 다시 리뷰 요청 보내주세요~!

> 완벽히 하지 않던 공부 습관이 들어나는 것 같아요.

완벽할 필요는 없죠. (뭐 저도 완벽한 습관을 가진 사람은 아닙니다.)
대신 리뷰에 공감이 갔다면 모든 파일을 보면서 꼼꼼히 보고, 이후 제출하기 전에 자신의 코드에 문제가 없는지 PR로 확인하는 습관을 가지면 좋을 것 같습니다.

llffmmgg committed:
refacotor: 애플리케이션 실행환경과 포트 분리, RestAssured에 port 설정

llffmmgg committed:
refactor: 클래스 내 코드 순서 수정

제대로 보자..

llffmmgg committed:
refactor: final -> static final 로 변경

llffmmgg reviewed (commented):
마지막 커멘트 남겨요 ! 자세한 리뷰 감사합니다 ㅠㅠㅠ

llffmmgg committed:
refactor: 지역변수에서 final 키워드 제거

leegwichan reviewed (approved):
긴 호흡을 가져가며 리뷰해 보았는데, 빠르게 답변해주셔서 감사합니다!
이제 step 7으로 넘어가보죠! 다음 리뷰에서 뵙겠습니다!
